rust/xaynet-server/src/storage/s3.rs

```diff
@@ -0,0 +1,290 @@
+use crate::settings::{S3BucketsSettings, S3Settings};
+use rusoto_core::{credential::StaticProvider, request::TlsError, HttpClient, RusotoError};
+use rusoto_s3::{
+    CreateBucketError,
+    CreateBucketOutput,
+    CreateBucketRequest,
+    DeleteObjectsError,
+    ListObjectsV2Error,
+    PutObjectError,
+    PutObjectOutput,
+    PutObjectRequest,
+    S3Client,
+    StreamingBody,
+    S3,
+};
+use std::sync::Arc;
+use thiserror::Error;
+
+use xaynet_core::mask::Model;
+
+type S3Result<T> = Result<T, S3Error>;
+
+#[derive(Debug, Error)]
+pub enum S3Error {
+    #[error("upload error: {0}")]
+    Upload(#[from] RusotoError<PutObjectError>),
+    #[error("create bucket error: {0}")]
+    CreateBucket(#[from] RusotoError<CreateBucketError>),
+    #[error("list objects error: {0}")]
+    ListObjects(#[from] RusotoError<ListObjectsV2Error>),
+    #[error("delete objects error: {0}")]
+    DeleteObjects(#[from] RusotoError<DeleteObjectsError>),
+    #[error("serialization failed")]
+    Serialization(#[from] bincode::Error),
+    #[error("empty response error")]
+    EmptyResponse,
+    #[error(transparent)]
+    HttpClient(#[from] TlsError),
+}
+
+#[derive(Clone)]
+pub struct Client {
+    buckets: Arc<S3BucketsSettings>,
+    s3_client: S3Client,
+}
+
+impl Client {
+    /// Creates a new S3 client. The client creates and maintains one bucket for storing global models.
+    ///
+    /// To connect to AWS-compatible services such as Minio, you need to specify a custom region.
+    /// ```
+    /// use rusoto_core::Region;
+    /// use xaynet_server::{
+    ///     settings::{S3BucketsSettings, S3Settings},
+    ///     storage::s3::Client,
+    /// };
+    ///
+    /// let region = Region::Custom {
+    ///     name: String::from("minio"),
+    ///     endpoint: String::from("http://127.0.0.1:9000"), // URL of minio
+    /// };
+    ///
+    /// let s3_settings = S3Settings {
+    ///     region,
+    ///     access_key: String::from("minio"),
+    ///     secret_access_key: String::from("minio123"),
+    ///     buckets: S3BucketsSettings {
+    ///         global_models: String::from("global-models"),
+    ///     },
+    /// };
+    ///
+    /// let store = Client::new(s3_settings);
+    /// ```
+    pub fn new(settings: S3Settings) -> S3Result<Self> {
+        let credentials_provider =
+            StaticProvider::new_minimal(settings.access_key, settings.secret_access_key);
+
+        let dispatcher = HttpClient::new()?;
+        Ok(Self {
+            buckets: Arc::new(settings.buckets),
+            s3_client: S3Client::new_with(dispatcher, credentials_provider, settings.region),
+        })
+    }
+
+    /// Uploads a global model.
+    pub async fn upload_global_model(&self, key: &str, global_model: &Model) -> S3Result<()> {
+        // As key for the global model we use the round_id and the seed of the round in which the
+        // global model was created.
```

Comment:
should this be part of the docstring? since the caller wouldn't otherwise know this (or is this just for the test? In which case maybe this comment belongs there...).

Reply:
you are right, I will move this line to the place where we store the model (in the unmasking phase) in #567 
After the restoring story I will add some module documentation to the s3/redis client.